crates/storage/src/connection/transaction.rs

```diff
@@ -10,325 +10,172 @@ use super::{EventsForBlock, TransactionDataForBlock, TransactionWithReceipt};
+            )
+            .context("Preparing insert transaction hash statement")?;
 
-            self.inner().execute(r"INSERT OR REPLACE INTO starknet_transactions (hash,  idx,  block_number,  tx,  receipt,  events) 
-                                                                        VALUES (:hash, :idx, :block_number, :tx, :receipt, :events)",
-            named_params![
-                ":hash": &transaction.hash(),
-                ":idx": &i.try_into_sql_int()?,
+        for (transaction, ..) in transactions.iter() {
+            insert_transaction_hash_stmt.execute(named_params![
+                ":hash": &transaction.hash,
                 ":block_number": &block_number,
-                ":tx": &tx_data,
-                ":receipt": &serialized_receipt,
-                ":events": &serialized_events,
-            ]).context("Inserting transaction data")?;
+            ])?;
         }
 
-        let events = transaction_data
-            .iter()
-            .filter_map(|data| data.events.as_ref())
-            .flatten();
-        self.insert_block_events(block_number, events)
-            .context("Inserting events into Bloom filter")?;
-        Ok(())
-    }
-
-    pub fn update_receipt(
-        &self,
-        block_number: BlockNumber,
-        transaction_idx: usize,
-        receipt: &Receipt,
-    ) -> anyhow::Result<()> {
         let mut compressor = zstd::bulk::Compressor::new(10).context("Create zstd compressor")?;
-        let receipt = dto::Receipt::from(receipt);
-        let serialized_receipt =
-            bincode::serde::encode_to_vec(&receipt, bincode::config::standard())
-                .context("Serializing receipt")?;
-        let serialized_receipt = compressor
-            .compress(&serialized_receipt)
-            .context("Compressing receipt")?;
-
-        let execution_status = match receipt {
-            dto::Receipt::V0(dto::ReceiptV0 {
-                execution_status: dto::ExecutionStatus::Succeeded,
-                ..
-            }) => 0,
-            dto::Receipt::V0(dto::ReceiptV0 {
-                execution_status: dto::ExecutionStatus::Reverted { .. },
-                ..
-            }) => 1,
+        let transactions: Vec<_> = transactions
+            .iter()
+            .map(|(transaction, receipt)| dto::TransactionWithReceipt {
+                transaction: dto::Transaction::from(transaction),
+                receipt: receipt.into(),
+            })
+            .collect();
+        let transactions = bincode::serde::encode_to_vec(transactions, bincode::config::standard())
+            .context("Serializing transaction")?;
+        let transactions = compressor
+            .compress(&transactions)
+            .context("Compressing transaction")?;
+        let encoded_events = match events {
+            Some(events) => {
+                let events: Vec<_> = events
+                    .iter()
+                    .map(|events| dto::Events::V0 {
```

Comment:
Shouldn't `dto::Events` contain Vec<Vec<Event>> so that the V0 tagging appears only once in the blob?

I think the point of this enum was that without this we cannot possibly make changes to the serialized event format. But if that's true we don't need to tag each transactions's events individually, right?